src/Illuminate/Collections/Collection.php

```diff
@@ -555,18 +555,37 @@ public function groupBy($groupBy, $preserveKeys = false)
      */
     public function keyBy($keyBy)
     {
-        $keyBy = $this->valueRetriever($keyBy);
-
         $results = [];
+        $items = $this->items;
 
-        foreach ($this->items as $key => $item) {
-            $resolvedKey = $keyBy($item, $key);
+        if (is_callable($keyBy) && ! is_string($keyBy)) {
+            $getter = $keyBy;
+        } elseif (is_string($keyBy) && ! str_contains($keyBy, '.')) {
+            $prop = $keyBy;
+            $getter = function ($item) use ($prop) {
+                if (is_array($item) && isset($item[$prop])) {
```

Comment:
Instead of `is_array()`, shouldn't we be using `Arr::accessible()` like in `data_get()`?

I wonder what is the performance difference of these changes on a collection of eloquent models.